In [1]:
import pandas as pd
import numpy as np
import glob
import os

# Set display options for easier viewing
pd.set_option('display.max_columns', 50)
pd.set_option('display.width', 150)

print("Current working directory:", os.getcwd())
print("\nFiles here:")
for f in os.listdir(os.getcwd()):
    print(f)

Current working directory: /Users/ruchapatwardhan/Desktop/IDX Internship

Files here:
.DS_Store
listings.ipynb
crmls_sold.py
listings_flagged.csv
CRMLSListing202602.csv
sold-1.ipynb
datascience.ipynb
sold_cleaned.csv
sold_final.csv
listings_final.csv
.gitignore
sold_combined.csv
listings_featured.csv
sold_flagged.csv
CRMLSSold202602.csv
listings_cleaned.csv
listings_combined.csv
.git
sold_featured.csv
crmls_listed.py
raw


In [2]:
with open("crmls_sold.py") as f:
    print(f.read())


import csv
import requests
from datetime import datetime

# Define the API endpoint
url = 'https://api-trestle.corelogic.com/trestle/odata/Property'

# Get token from IDX Exchange secure proxy instead of exposing CoreLogic credentials
auth_endpoint = 'https://idxexchange.com/internal-api/trestle_token.php?key=IDXEXCHANGE2026_CHANGE_THIS'

response = requests.get(auth_endpoint)
response.raise_for_status()

# Parse the response to extract the token

token = response.json().get('access_token')

if token:
    # Define the headers with the token
    headers = {
        'Authorization': f'Bearer {token}'
    }

    # Define the parameters for the API request
    params = {
        '$select': 'BuyerAgentAOR, ListAgentAOR, Flooring,ViewYN,WaterfrontYN,BasementYN,PoolPrivateYN, OriginalListPrice, ListingKey,ListAgentEmail,CloseDate,ClosePrice,ListAgentFirstName,ListAgentLastName,Latitude,Longitude,UnparsedAddress,PropertyType,LivingArea,ListPrice,DaysOnMarket,ListOfficeName,BuyerOfficeName,CoL

In [3]:
df_final = pd.read_csv("sold_final.csv")
print("Shape:", df_final.shape)
print("\nColumns:", df_final.columns.tolist())
df_final.head()

/var/folders/pn/gnz_53jn6wxbzf_7c8wklgf40000gn/T/ipykernel_2714/1897815327.py:1: DtypeWarning: Columns (0,1,4,70,71,72,74,75) have mixed types. Specify dtype option on import or set low_memory=False.
  df_final = pd.read_csv("sold_final.csv")


Shape: (312985, 97)

Columns: ['BuyerAgentAOR', 'ListAgentAOR', 'Flooring', 'ViewYN', 'WaterfrontYN', 'BasementYN', 'PoolPrivateYN', 'OriginalListPrice', 'ListingKey', 'ListAgentEmail', 'CloseDate', 'ClosePrice', 'ListAgentFirstName', 'ListAgentLastName', 'Latitude', 'Longitude', 'UnparsedAddress', 'PropertyType', 'LivingArea', 'ListPrice', 'DaysOnMarket', 'ListOfficeName', 'BuyerOfficeName', 'CoListOfficeName', 'ListAgentFullName', 'CoListAgentFirstName', 'CoListAgentLastName', 'BuyerAgentMlsId', 'BuyerAgentFirstName', 'BuyerAgentLastName', 'AssociationFeeFrequency', 'ListingKeyNumeric', 'MLSAreaMajor', 'CountyOrParish', 'MlsStatus', 'ElementarySchool', 'AttachedGarageYN', 'ParkingTotal', 'BuilderName', 'PropertySubType', 'LotSizeAcres', 'SubdivisionName', 'BuyerOfficeAOR', 'YearBuilt', 'StreetNumberNumeric', 'ListingId', 'BathroomsTotalInteger', 'City', 'BuildingAreaTotal', 'BedroomsTotal', 'ContractStatusChangeDate', 'CoBuyerAgentFirstName', 'PurchaseContractDate', 'ListingContractD

,BuyerAgentAOR,ListAgentAOR,Flooring,ViewYN,WaterfrontYN,BasementYN,PoolPrivateYN,OriginalListPrice,ListingKey,ListAgentEmail,CloseDate,ClosePrice,ListAgentFirstName,ListAgentLastName,Latitude,Longitude,UnparsedAddress,PropertyType,LivingArea,ListPrice,DaysOnMarket,ListOfficeName,BuyerOfficeName,CoListOfficeName,ListAgentFullName,...,BuyerAgencyCompensationType,BuyerAgencyCompensation,latfilled,lonfilled,year_month,rate_30yr_fixed,invalid_close_price,invalid_living_area,invalid_days_on_market,listing_after_close_flag,purchase_after_close_flag,negative_timeline_flag,missing_coords,zero_coords,wrong_longitude,price_ratio,price_per_sqft,year,month,yr_mo,listing_to_contract_days,contract_to_close_days,outlier_close_price,outlier_living_area,outlier_days_on_market
0,SanDiego,SanDiego,NaN,False,NaN,NaN,False,759900.0,522107581,mdarwich12@gmail.com,2024-01-05,815000.0,Michael,Darwich,NaN,NaN,2811 C Avenue,Residential,1974.0,759900.0,33,Berkshire Hathaway HomeServices California Pro...,FOSTER HAMILTON Real Estate,NaN,Michael Darwich,...,NaN,NaN,NaN,NaN,2024-01,6.6425,False,False,False,False,False,False,True,False,False,1.07251,412.867275,2024,1,2024-01,114.0,919.0,False,False,False
1,Mlslistings,Mlslistings,NaN,False,NaN,NaN,NaN,NaN,1079166779,davidmartz@compass.com,2024-01-30,858000.0,David,Martz,NaN,NaN,2199 N Berne Drive,Residential,1995.0,858000.0,0,Compass,NaN,NaN,David Martz,...,NaN,NaN,NaN,NaN,2024-01,6.6425,False,False,False,False,True,False,True,False,False,NaN,430.075188,2024,1,2024-01,188.0,-188.0,False,False,False
2,Southland,Southland,NaN,False,NaN,NaN,False,1890500.0,1075037759,karen.klein@theagencyre.com,2024-01-29,1890500.0,Karen,Klein,34.164376,-118.558749,19524 Winifred Street,Residential,3194.0,1890500.0,0,The Agency,The Agency - Studio City,NaN,Karen Klein,...,NaN,NaN,NaN,NaN,2024-01,6.6425,False,False,False,False,False,False,False,False,False,1.00000,591.891046,2024,1,2024-01,0.0,0.0,False,False,False
3,CoastalMendocino,CoastalMendocino,Tile,True,NaN,NaN,False,1950000.0,1063453216,kira@mendosir.com,2024-01-22,1950000.0,Kira,Meade,39.419080,-123.814842,31451 Bay View Avenue,Residential,2100.0,1950000.0,-36,Mendo Sotheby's International Realty,West Coast Property Management and Sales,NaN,Kira Meade,...,NaN,NaN,NaN,NaN,2024-01,6.6425,False,False,True,False,False,False,False,False,False,1.00000,928.571429,2024,1,2024-01,32.0,34.0,False,False,False
4,Mlslistings,Mlslistings,"Tile,Wood",True,NaN,NaN,NaN,NaN,1061988701,dannyduong@gmail.com,2024-01-31,2340000.0,Danny,Duong,NaN,NaN,3114 Rasmus Circle,Residential,2442.0,2340000.0,0,Infiniti Real Estate,Out of Area Office,NaN,Danny Duong,...,NaN,NaN,NaN,NaN,2024-01,6.6425,False,False,False,False,True,False,True,False,False,NaN,958.230958,2024,1,2024-01,33.0,-33.0,False,False,False


In [4]:
print(df_final['PropertySubType'].value_counts())
print("\nDate range:", df_final['CloseDate'].min(), "to", df_final['CloseDate'].max())

PropertySubType
SingleFamilyResidence    228992
Condominium               54978
Townhouse                 20487
ManufacturedOnLand         3993
Duplex                     1608
StockCooperative           1400
Cabin                       307
Triplex                     211
MixedUse                    111
Quadruplex                   61
BoatSlip                     60
OwnYourOwn                   42
ManufacturedHome             30
MobileHome                   29
Loft                         19
CoOwnership                  14
Timeshare                    11
Studio                        9
Farm                          8
DeededParking                 6
Name: count, dtype: int64

Date range: 2024-01-01 to 2026-01-31


In [5]:
# Filter to the required scope
df = df_final[df_final['PropertySubType'] == 'SingleFamilyResidence'].copy()
print(f"Rows before filter: {len(df_final)}")
print(f"Rows after filter: {len(df)}")

# Convert CloseDate to datetime for time based splitting
df['CloseDate'] = pd.to_datetime(df['CloseDate'])

# Use the most recent month as test, prior months as train
df['close_year_month'] = df['CloseDate'].dt.to_period('M')
most_recent_month = df['close_year_month'].max()
print("\nMost recent month (test set):", most_recent_month)

test_df = df[df['close_year_month'] == most_recent_month].copy()

# Try a training window of X months before the test month, start with X = 6
X = 6
train_months = pd.period_range(end=most_recent_month - 1, periods=X, freq='M')
train_df = df[df['close_year_month'].isin(train_months)].copy()

print(f"\nTrain set (X={X} months): {len(train_df)} rows, {train_df['close_year_month'].min()} to {train_df['close_year_month'].max()}")
print(f"Test set: {len(test_df)} rows")

Rows before filter: 312985
Rows after filter: 228992

Most recent month (test set): 2026-01

Train set (X=6 months): 55024 rows, 2025-07 to 2025-12
Test set: 5867 rows


Week 4

In [6]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

# Pick a simple set of numeric features for the baseline
features = ['LivingArea', 'BedroomsTotal', 'BathroomsTotalInteger', 'LotSizeSquareFeet', 'YearBuilt']
target = 'ClosePrice'

# Drop rows with missing values in these columns for the baseline
train_model_df = train_df[features + [target]].dropna()
test_model_df = test_df[features + [target]].dropna()

print(f"Train rows after dropping missing: {len(train_model_df)}")
print(f"Test rows after dropping missing: {len(test_model_df)}")

X_train = train_model_df[features]
y_train = train_model_df[target]
X_test = test_model_df[features]
y_test = test_model_df[target]

# Train baseline linear regression
model = LinearRegression()
model.fit(X_train, y_train)

# Evaluate
train_preds = model.predict(X_train)
test_preds = model.predict(X_test)

train_r2 = r2_score(y_train, train_preds)
test_r2 = r2_score(y_test, test_preds)

print(f"\nTrain R²: {train_r2:.4f}")
print(f"Test R²: {test_r2:.4f}")

# Look at coefficients to understand what's driving predictions
coef_df = pd.DataFrame({'Feature': features, 'Coefficient': model.coef_})
print("\nCoefficients:")
print(coef_df)

Train rows after dropping missing: 54055
Test rows after dropping missing: 5737

Train R²: 0.3070
Test R²: 0.3076

Coefficients:
                 Feature    Coefficient
0             LivingArea     346.735945
1          BedroomsTotal  -29138.941694
2  BathroomsTotalInteger  122232.012157
3      LotSizeSquareFeet      -0.000439
4              YearBuilt   -7128.386378


Baseline Linear Regression, Test R² = 0.3076
Features used: LivingArea, BedroomsTotal, BathroomsTotalInteger, LotSizeSquareFeet, YearBuilt
~1,000 train rows and ~130 test rows dropped due to missing values in these features
Negative coefficients on Bedrooms and YearBuilt likely reflect multicollinearity, not causal relationships

week 5

In [7]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor

# Decision Tree
tree_model = DecisionTreeRegressor(max_depth=10, random_state=42)
tree_model.fit(X_train, y_train)

tree_train_preds = tree_model.predict(X_train)
tree_test_preds = tree_model.predict(X_test)

tree_train_r2 = r2_score(y_train, tree_train_preds)
tree_test_r2 = r2_score(y_test, tree_test_preds)

print("Decision Tree:")
print(f"Train R²: {tree_train_r2:.4f}")
print(f"Test R²: {tree_test_r2:.4f}")

# Random Forest
rf_model = RandomForestRegressor(n_estimators=100, max_depth=10, random_state=42, n_jobs=-1)
rf_model.fit(X_train, y_train)

rf_train_preds = rf_model.predict(X_train)
rf_test_preds = rf_model.predict(X_test)

rf_train_r2 = r2_score(y_train, rf_train_preds)
rf_test_r2 = r2_score(y_test, rf_test_preds)

print("\nRandom Forest:")
print(f"Train R²: {rf_train_r2:.4f}")
print(f"Test R²: {rf_test_r2:.4f}")

# Compare all three models side by side
comparison = pd.DataFrame({
    'Model': ['Linear Regression', 'Decision Tree', 'Random Forest'],
    'Train R²': [train_r2, tree_train_r2, rf_train_r2],
    'Test R²': [test_r2, tree_test_r2, rf_test_r2]
})
print("\nModel Comparison:")
print(comparison)

# Feature importance from Random Forest
importance_df = pd.DataFrame({
    'Feature': features,
    'Importance': rf_model.feature_importances_
}).sort_values('Importance', ascending=False)
print("\nRandom Forest Feature Importance:")
print(importance_df)

Decision Tree:
Train R²: 0.4496
Test R²: 0.3510

Random Forest:
Train R²: 0.4796
Test R²: 0.4138

Model Comparison:
               Model  Train R²   Test R²
0  Linear Regression  0.307021  0.307634
1      Decision Tree  0.449632  0.350967
2      Random Forest  0.479615  0.413799

Random Forest Feature Importance:
                 Feature  Importance
0             LivingArea    0.492998
4              YearBuilt    0.279309
3      LotSizeSquareFeet    0.140074
2  BathroomsTotalInteger    0.061050
1          BedroomsTotal    0.026569
